src/storage/v2/schema_validator.cpp

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 Memgraph Ltd.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt; by using this file, you agree to be bound by the terms of the Business Source
+// License, and you may not use this file except in compliance with the Business Source License.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+#include "storage/v2/schema_validator.hpp"
+
+#include <bits/ranges_algo.h>
+#include <cstddef>
+#include <ranges>
+
+#include "storage/v2/schemas.hpp"
+
+namespace memgraph::storage {
+
+bool operator==(const SchemaViolation &lhs, const SchemaViolation &rhs) {
+  return lhs.status == rhs.status && lhs.label == rhs.label &&
+         lhs.violated_schema_property == rhs.violated_schema_property &&
+         lhs.violated_property_value == rhs.violated_property_value;
+}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label) : status{status}, label{label} {}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label, SchemaProperty violated_schema_property)
+    : status{status}, label{label}, violated_schema_property{violated_schema_property} {}
+
+SchemaViolation::SchemaViolation(ValidationStatus status, LabelId label, SchemaProperty violated_schema_property,
+                                 PropertyValue violated_property_value)
+    : status{status},
+      label{label},
+      violated_schema_property{violated_schema_property},
+      violated_property_value{violated_property_value} {}
+
+SchemaValidator::SchemaValidator(Schemas &schemas) : schemas_{schemas} {}
+
+[[nodiscard]] std::optional<SchemaViolation> SchemaValidator::ValidateVertexCreate(
+    LabelId primary_label, const std::vector<LabelId> &labels,
+    const std::vector<std::pair<PropertyId, PropertyValue>> &properties) const {
+  // Schema on primary label
+  const auto schema = schemas_.GetSchema(primary_label);
```

Comment:
As `Schemas::GetSchema` is called in the hot path and it contains a vector, consider using a pointer instead. If the function returns a `nullptr`, it means the schema doesn't exists. This way you can avoid copying the vector inside the schema.